deno_typescript/ops.rs

```diff
@@ -0,0 +1,160 @@
+use crate::get_asset;
+use crate::TSState;
+use deno::CoreOp;
+use deno::ErrBox;
+use deno::ModuleSpecifier;
+use deno::Op;
+use deno::OpId;
+use serde::Deserialize;
+use serde_json::json;
+use serde_json::Value;
+
+#[derive(Debug)]
+pub struct WrittenFile {
+  pub url: String,
+  pub module_name: String,
+  pub source_code: String,
+}
+
+fn dispatch2(
+  s: &mut TSState,
+  op_id: OpId,
+  control_buf: &[u8],
+) -> Result<Value, ErrBox> {
+  let v = serde_json::from_slice(control_buf)?;
+  match op_id {
+    49 => read_file(s, v),
+    50 => exit(s, v),
+    51 => write_file(s, v),
+    52 => resolve_module_names(s, v),
+    53 => set_emit_result(s, v),
+    _ => unreachable!(),
+  }
+}
+
+pub fn dispatch_op(s: &mut TSState, op_id: OpId, control_buf: &[u8]) -> CoreOp {
+  let result = dispatch2(s, op_id, control_buf);
+  let response = match result {
+    Ok(v) => json!({ "ok": v }),
+    Err(err) => json!({ "err": err.to_string() }),
+  };
+  let x = serde_json::to_string(&response).unwrap();
+  let vec = x.to_string().into_bytes();
+  Op::Sync(vec.into_boxed_slice())
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ReadFile {
+  file_name: String,
+  language_version: Option<i32>,
+  should_create_new_source_file: bool,
+}
+
+fn read_file(_s: &mut TSState, v: Value) -> Result<Value, ErrBox> {
+  let v: ReadFile = serde_json::from_value(v)?;
+  let (module_name, source_code) = if v.file_name.starts_with("$asset$/") {
+    let asset = v.file_name.replace("$asset$/", "");
+    // TODO(ry) remove unwrap below, return error "no such asset"
```

Comment:
io::Error::new(io::ErrorKind::NotFound, "Asset not found");

Reply:
Done